MSBuild.exe crashes when specifying multiple target names

### Steps to reproduce

```
msbuild.exe /t:Foo /t:Bar
```

### Expected  behavior
Build `Foo` and `Bar`

### Actual behavior
MSBuild crashes

```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
System.InvalidOperationException: Sequence contains more than one element
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Boolean needToValidateProject, String schemaFile, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler)
```
This is a regression caused by https://github.com/Microsoft/msbuild/pull/3278

--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2002,6 +2002,27 @@ namespace Microsoft.Build.UnitTests
             logContents.ShouldContain(guid2);
         }
 
+        /// <summary>
+        /// We check if there is only one target name specified and this logic caused a regression: https://github.com/Microsoft/msbuild/issues/3317
+        /// </summary>
+        [Fact]
+        public void MultipleTargetsDoesNotCrash()
+        {
+            string projectContents = ObjectModelHelpers.CleanupFileContents($@"<Project xmlns=""http://schemas.microsoft.com/developer/msbuild/2003"">
+  <Target Name=""Target1"">
+    <Message Text=""7514CB1641A948D0A3930C5EC2DC1940"" />
+  </Target>
+  <Target Name=""Target2"">
+    <Message Text=""E2C73B5843F94B63B067D9BEB2C4EC52"" />
+  </Target>
+</Project>");
+
+            string logContents = ExecuteMSBuildExeExpectSuccess(projectContents, arguments: "/t:Target1 /t:Target2");
+
+            logContents.ShouldContain("7514CB1641A948D0A3930C5EC2DC1940", () => logContents);
+            logContents.ShouldContain("E2C73B5843F94B63B067D9BEB2C4EC52", () => logContents);
+        }
+
         private string CopyMSBuild()
         {
             string dest = null;
@@ -2066,7 +2087,7 @@ namespace Microsoft.Build.UnitTests
 
                 string output = RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" {String.Join(" ", arguments)}", out success, _output);
 
-                success.ShouldBeTrue();
+                success.ShouldBeTrue(() => output);
 
                 return output;
             }

--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1118,7 +1118,7 @@ namespace Microsoft.Build.CommandLine
                         try
                         {
                             // Determine if the user specified /Target:Restore which means we should only execute a restore in the fancy way that /restore is executed
-                            bool restoreOnly = String.Equals(request.TargetNames.SingleOrDefault(), MSBuildConstants.RestoreTargetName, StringComparison.OrdinalIgnoreCase);
+                            bool restoreOnly = request.TargetNames.Count == 1 && String.Equals(request.TargetNames.First(), MSBuildConstants.RestoreTargetName, StringComparison.OrdinalIgnoreCase);
 
                             if (enableRestore || restoreOnly)
                             {

